keanu-project/src/main/java/io/improbable/keanu/util/dot/WriteDot.java

```diff
@@ -0,0 +1,279 @@
+     *
+     * @param fileName name of the output file.
+     * @param net network to be written out to DOT format.
+     */
+    public static void outputDot(String fileName, BayesianNetwork net) {
+
+        // Get any vertex in the network and set the degree to infinity to print out the entire network.
+        if (net.getAllVertices().size() == 0) return;
+        Vertex someVertex = net.getAllVertices().get(0);
+        outputDot(fileName, someVertex, Integer.MAX_VALUE);
+    }
+
+
+    /**
+     * Outputs the given graph to a DOT file which can be used by various graph visualisers to generate a visual representation of the graph.
+     * Read more about DOT format here: https://en.wikipedia.org/wiki/DOT_(graph_description_language)
+     *
+     * @param fileName name of the output file
+     * @param vertex vertex around which the graph will be visualised
+     * @param degree degree of connections to be visualised; for instance, if the degree is 1,
+     *               only connections between the vertex amd it's parents and children will be written out to the DOT file.
+     */
+    public static void outputDot(String fileName, Vertex vertex, int degree) {
+        try {
+            // Reset variables
+            idsToLabelStrings = new HashMap<>();
+            verticeIdsToConnectionInfo = new HashMap<>();
+
+            File outputFile = getOutputFile(fileName);
+
+            writer = new BufferedWriter(new FileWriter(outputFile));
+            writer.write("digraph BayesianNetwork {\n");
+
+            // Iterate over vertices and obtain the necessary label and conneciton info.
+            obtainGraphInfo(vertex, degree);
+
+            // Write out labels and connections.
+            outputConnections(verticeIdsToConnectionInfo.values());
+            outputLabels();
+
+            writer.write("}");
+            writer.close();
+        }
+        catch (IOException e) {
+            System.out.println("Encountered an issue writing to the specified file (" + fileName + "):");
+            e.printStackTrace();
+        }
+    }
+
+
+    // Make sure the output directory exists and that no files are being overwritten.
+    private static File getOutputFile(String fileName) {
+
+        File outputFile = new File(fileName);
+
+        // Make sure the output directory exists.
+        if (outputFile.getParentFile() != null && !outputFile.getParentFile().exists()){
+            outputFile.getParentFile().mkdirs();
+        }
+
+        // Add an index to the filename if a file with the specified name already exists.
+        if (outputFile.exists()) {
+            String baseName = FilenameUtils.getBaseName(fileName);
+            String extension = FilenameUtils.getExtension(fileName);
+            int counter = 1;
+            while(outputFile.exists())
+            {
+                outputFile = new File(outputFile.getParent(), baseName + "_" + (counter++) + "." + extension);
```

Comment:
Do we want to do 1000s of Loops if lots of these files exist?  Should we put an upper limit on it?

Also Nit: I'm not a fan of expressions with side effects within a single line of code (I think it's worth being explicit about every change you make to program state).

Reply:
Do we expect the user to have created and be storing thousands of such files? I'd be reluctant to put an upper limit to it, as I think it's never nice to overwrite a file without an explicit permission from the user.

Alternatively, we could timestamp them, but that looks a bit ugly. 

If you strongly think we should have an upper limit, I'll change it.

Reply:
Re making it cleaner - would you prefer something like the following?

``` java
String newFileName = baseName + "_" + (counter++) + "." + extension;
outputFile = new File(outputFile.getParent(), newFileName);```